include/picongpu/simulationControl/DomainAdjuster.hpp

```diff
@@ -0,0 +1,372 @@
+                    std::cout << "Dimension " << dimNames[ dim ] <<
+                        ": local grid size must be equal for all devices because moving window is enabled. Adjust from " <<
+                        m_localDomainSize[ dim ] << " to " <<  globalMax << std::endl;
+                }
+
+                m_localDomainSize[ dim ] = globalMax;
+            }
+        }
+
+        /** derive the global domain size
+         *
+         * Calculate the global domain size.
+         *
+         * @param dim dimension to update
+         */
+        void deriveGlobalDomainSize( size_t const dim )
+        {
+            uint64_t validGlobalGridSize = 0;
+
+            deriveLocalDomainSize( dim );
+
+            if( m_movingWindowEnabled && dim == 1 )
+            {
+                // the local sizes in slide direction must be equal sized
+                validGlobalGridSize = static_cast< uint64_t >( m_localDomainSize[ dim ] * m_numDevices[ dim ] );
+            }
+            else
+            {
+                uint64_t localDomainSize = static_cast< uint64_t >( m_localDomainSize[ dim ] );
+                pmacc::mpi::MPIReduce mpiReduce;
+                mpiReduce(
+                    pmacc::nvidia::functors::Add(),
+                    &validGlobalGridSize,
+                    &localDomainSize,
+                    1,
+                    pmacc::mpi::reduceMethods::AllReduce()
+                );
+                /* since we are not doing independent reduces per slice we need
+                 * to adjust the reduce result by dividing the sizes of all other dimensions
+                 * we are not check within the method call
+                 */
+                validGlobalGridSize /= static_cast< uint64_t >(
+                    m_numDevices.productOfComponents() / m_numDevices[ dim ]
+                );
+
+            }
+
+            if( m_isMaster && validGlobalGridSize != static_cast< uint64_t >( m_globalDomainSize[ dim ] ) )
+            {
+                    std::cout << "Dimension " << dimNames[ dim ] <<
+                        ": global grid size adjusted from " <<
+                        m_globalDomainSize[ dim ] << " to " <<  validGlobalGridSize << std::endl;
+            }
+
+            m_globalDomainSize[ dim ] = static_cast< int >( validGlobalGridSize );
+        }
+
+        DataSpace< simDim > m_globalDomainSize;
+        DataSpace< simDim > m_localDomainSize;
+        DataSpace< simDim > m_localDomainOffset;
+        DataSpace< simDim > const m_numDevices;
+        DataSpace< simDim > const m_mpiPosition;
+        DataSpace< simDim > const m_isPeriodic;
+        bool const m_movingWindowEnabled;
+        bool const m_isMaster;
+        /**! lookup table to translate a dimension index into a name
+         *
+         * \warning `= { { ... } }` is not required by the c++11 standard but
+         * is necessary for g++ 4.9
+         */
+        std::array< char, 3 > dimNames = { { 'x', 'y', 'z' } };
```

Comment:
NOTE: g++ 4.9 is not standard conform and requires extra brackets around the initializer list.

Reply:
I think that's actually a C++11 thing and was lifted in C++14, as I read the other day on SO.

Reply:
NO as I know the double brackets was only needed if you have not the assign operator but since gcc 4.9 was early stage of C++11 it is maybe a bug but not really important.